hudi-common/src/test/java/org/apache/hudi/common/util/hash/TestBucketIndexUtil.java

```diff
@@ -0,0 +1,146 @@
+    argsList.add(Arguments.of(400, 1000));
+    argsList.add(Arguments.of(401, 1000));
+
+    return argsList.stream();
+  }
+
+  private static Stream<Arguments> noPartitionParams() {
+    List<Arguments> argsList = new ArrayList<>();
+
+    argsList.add(Arguments.of(10, 50));
+    argsList.add(Arguments.of(11, 50));
+    argsList.add(Arguments.of(100, 50));
+    argsList.add(Arguments.of(101, 50));
+
+    return argsList.stream();
+  }
+
+  @ParameterizedTest
+  @MethodSource("partitionParams")
+  void testPartition(int parallelism, int bucketNumber) {
+    Map<Integer, Integer> parallelism2TaskCount = new HashMap<>();
+    final Functions.Function2<String, Integer, Integer> partitionIndexFunc =
+        BucketIndexUtil.getPartitionIndexFunc(bucketNumber, parallelism);
+    initPartitionData(parallelism2TaskCount, bucketNumber, partitionIndexFunc);
+    checkResult(parallelism2TaskCount, parallelism, bucketNumber);
+  }
+
+  @ParameterizedTest
+  @MethodSource("noPartitionParams")
+  void testNoPartition(int parallelism, int bucketNumber) {
+    Map<Integer, Integer> parallelism2TaskCount = new HashMap<>();
+    final Functions.Function2<String, Integer, Integer> partitionIndexFunc =
+        BucketIndexUtil.getPartitionIndexFunc(bucketNumber, parallelism);
+    initNoPartitionData(parallelism2TaskCount, bucketNumber, partitionIndexFunc);
+    checkResult(parallelism2TaskCount, parallelism, bucketNumber);
+  }
+
+  private static void putIndexCount(Map<Integer, Integer> parallelism2TaskCount, int workIndex) {
+    if (parallelism2TaskCount.containsKey(workIndex)) {
+      parallelism2TaskCount.put(workIndex, parallelism2TaskCount.get(workIndex) + 1);
+    } else {
+      parallelism2TaskCount.put(workIndex, 1);
+    }
+  }
+
+  private void checkResult(Map<Integer, Integer> parallelism2TaskCount, int parallelism, int bucketNumber) {
+    int sum = 0;
+    for (int v : parallelism2TaskCount.values()) {
+      sum = sum + v;
+    }
+    int avg = sum / parallelism;
+    double minToleranceValue = avg * 0.8;
+    double maxToleranceValue = avg * 1.2;
+    final ArrayList<Integer> outOfLimit = new ArrayList<>();
+    final ArrayList<Integer> inLimit = new ArrayList<>();
+    for (int v : parallelism2TaskCount.values()) {
+      // if parallelism is too bigger, first condition will false although the diff is 1
+      // for example, avg is 4, v is 5 or 3 all will out of limit, so add second condition
+      if ((v >= minToleranceValue) && (v <= maxToleranceValue) || ((Math.abs(v - avg) <= 2))) {
+        inLimit.add(v);
+      } else {
+        outOfLimit.add(v);
+      }
+    }
+    assertEquals(0, outOfLimit.size());
+    // parallelism2TaskCount size need Infinitely close parallelism or (bucketNumber * 8)
+    if (parallelism >= (bucketNumber * 8)) {
+      assertTrue(parallelism2TaskCount.size() >= (bucketNumber * 8 * 0.9));
+    } else {
+      assertTrue(parallelism2TaskCount.size() >= parallelism * 0.9);
+    }
```

Comment:
add a new check, the work thread need infinitely close parallelism or (bucketNumber * partitionNumber) if we have enougth resource